hypothesis-python/src/hypothesis/strategies/_internal/core.py

```diff
@@ -1083,7 +1111,17 @@ def as_strategy(strat_or_callable, thing, final=True):
         # way to tell and we just have to assume that everything is required.
         # See https://github.com/python/cpython/pull/17214 for details.
         optional = getattr(thing, "__optional_keys__", ())
-        anns = {k: from_type(v) for k, v in get_type_hints(thing).items()}
+        anns = {}
+        for k, v in get_type_hints(thing).items():
+            if hasattr(v, "__origin__") and v.__origin__ in NotRequiredTypes:
+                optional = optional.union({k})
+                anns.update({k: from_type(v.__args__[0])})
```

Comment:
What happens if an attribute is e.g. `key: Required` (with no inner type)?  I'd like to give a better error message than the IndexError here.

Reply:
In my env (python 3.10) it will give `TypeError: Plain typing_extensions.NotRequired is not valid as type argument` I can add tests to make sure it is the case if you like it?

Reply:
Yep - while we *hope* that people will get useful feedback from other tools, we want to make sure that Hypothesis is as helpful as possible - if nothing else has worked, that's especially important!

(Hypothesis would be *so much simpler* if we ripped out all the code that exists only to give helpful errors, but also *so much harder* to learn and use.  I think it's worth the effort for maintainers, in the end, especially for tricky testing tools.)

Reply:
Ok, I will see if I can write so test to see if it is the same across platforms. The thing is, I have added extra code to raise custom exceptions but this error message took priority. If this is the case across platforms (which means the error message I added will never show up) I would rather keep the code clean. What do you think? I have stash the work so I can add back easily :-)